lib/core/src/Cardano/Wallet.hs

```diff
@@ -1726,36 +1729,53 @@ joinStakePool
         , AddressIndexDerivationType k ~ 'Soft
         )
     => ctx
+    -> NetworkParameters
+    -> [PoolId]
+    -> PoolId
+    -> PoolLifeCycleStatus
     -> WalletId
-    -> (PoolId, [PoolId])
     -> ArgGenChange s
     -> Passphrase "raw"
     -> ExceptT ErrJoinStakePool IO (Tx, TxMeta, UTCTime)
-joinStakePool ctx wid (pid, pools) argGenChange pwd = db & \DBLayer{..} -> do
-    (isKeyReg, walMeta) <- mapExceptT atomically
-        $ withExceptT ErrJoinStakePoolNoSuchWallet
-        $ (,) <$> isStakeKeyRegistered (PrimaryKey wid)
-              <*> withNoSuchWallet wid (readWalletMeta (PrimaryKey wid))
+joinStakePool ctx np knownPools pid poolStatus wid argGenChange pwd =
+    db & \DBLayer{..} -> do
 
-    withExceptT ErrJoinStakePoolCannotJoin $ except $
-        guardJoin pools (walMeta ^. #delegation) pid
+        (isKeyReg, walMeta) <- mapExceptT atomically
+            $ withExceptT ErrJoinStakePoolNoSuchWallet
+            $ (,) <$> isStakeKeyRegistered (PrimaryKey wid)
+                  <*> withNoSuchWallet wid (readWalletMeta (PrimaryKey wid))
 
-    let action = if isKeyReg then Join pid else RegisterKeyAndJoin pid
-    liftIO $ traceWith tr $ MsgIsStakeKeyRegistered isKeyReg
+        currentTime <- liftIO getCurrentTime
+        let currentEpoch = view #epochNumber
+                $ fromMaybe W.slotMinBound
+                $ W.slotAt' sp currentTime
```

Comment:
I wasn't able to fully remove `slotAt'` yet in #1901, but we need to use the `TimeInterpreter` for this. Otherwise the calculation will be incorrect.

I suggest using `ongoingSlotAt`, `startTime`, and running the resulting `Qry` using the `timeInterpreter` field from `networkLayer`. Happy to help if that doesn't work out for some reason.

Reply:
> I wasn't able to fully remove `slotAt'` yet in #1901, but we need to use the `TimeInterpreter` for this. Otherwise the calculation will be incorrect.

Thanks for letting me know. When are you planning to remove the `slotAt'` function? The `getNetworkInformation` endpoint also uses this function in a similar way to determine the _next_ epoch. If that function is updated to use the new API, then we would presumably have to do something similar for `joinStakePool`?

Reply:
> If that function is updated to use the new API, then we would presumably have to do something similar for joinStakePool?

Yes, they should be similar.

Here you handle the `Nothing` case, where it means the current time is behind the genesis date, I think.

With the hard fork, there the query may (also) fail if the node is not in sync, if the current time is beyond the forecast range of the interpreter (think this is up to the next epoch, but not sure exactly)

Reply:
> > If that function is updated to use the new API, then we would presumably have to do something similar for joinStakePool?
> 
> Yes, they should be similar.
> 
> Here you handle the `Nothing` case, where it means the current time is behind the genesis date, I think.
> 
> With the hard fork, there the query may (also) fail if the node is not in sync, if the current time is beyond the forecast range of the interpreter (think this is up to the next epoch, but not sure exactly)

Hi @Anviking 

Thanks for your help with this!

I've (fingers crossed) fixed this with the following commits:

* 255210f8b5e6b58a7f4ce47ab83e26b954bc915b
* 7660579b295bd8ad78085da453c6c1c670f45bd5
* 18ef12a77a7a600738e52f85efe292de87168eef
* 23010f8ab79daa7f0ff86d0693b0f02b20c97c0a